### ISSUE TITLE
Add stubs for all exercises

### ISSUE THREAD

**Issue by iHiD** [exercism/exercism#4789]
Some tracks currently have stub files for exercises. Other tracks only provide the test file. I propose making stubs a requirement on all exercises on all tracks. My reasons are:

- **Consistency: ** A few drive for me is to make Exercism more consistent. This gives a better experience to the user and less confusion.
- **Helping beginners: ** Creating a file is something which requires someone to Google. They have to work out correct file extensions, they have to work out correct basic syntax, etc. For C# a basic solution to two fer involves discovering and working out how to arrange all the following terms `public`, `static`, `using`, `class` and `string`. For Ruby it requires `module` or `class` and `def` - neither of which would necessarily be clear. Providing stubs reduces this hard learning curve. 
- **Reduction in support:** We get emailed a lot by people having no idea what to do to get started and a stub file would dramatically reduce that barrier.
- **Adding progression:** Each core exercise can have a reduced stub, until we just have a blank file with the correct filename. This gives a student a bit more work to do each time, but also means they have previous exercises to refer back to.
- **Auto-mentoring: ** We need to be able to predict the file that someone will upload to be able to run static analysis on it. Having stubs enforces the file name of the solution, which reduces complexity for the auto-mentoring maintainers.

The two negatives I have heard to this proposal are:
- **Not teaching error messages:** If someone sees an error message of "file missing" or "require failed" and learn how to deal with it, then that is valuable. 
- **Learning how to create files:** Creating the file means you become familiar with the file extension etc.
- **Another job for maintainers:** All tracks will need to do this bit of housekeeping, and if we have progressively smaller stubs, then some thought needs to go into that.

My strong opinion is that the positives outway the negatives. I would also like to suggest adding the progression through reduced stubs into the Track Anatomy Project.

Have I missed any pros/cons? Does anyone have anything to add that I've not considered?

If people could :+1: and :-1: on this issue with their preference, I would appreciate it.

**Comment by SleeplessByte** [comment]
Sidenote:
- you've written *two (2)* negatives and are already listing three
- can you please fix the bolding? Remove the last space

In general I agree with this.

Another negative:
- you're going to need to make choices that might not be idiomatic but preference (`module` over `class` etc).

> enforces the file name of the solution, which reduces complexity for the auto-mentoring maintainers.
-  No, it makes it much more likely. I personally would go the way of the `javascript-analyzer`: it searches for `exercise.js` but if it can't be found it will use whatever `.js` it can find that is not a `spec` file.

**Comment by cmccandless** [comment]
Python has had stubs for a while now.

I can attest to the benefit of knowing where to start. Whenever I start a new track in a language I haven't worked with before, I like to have something to point me the in right direction so I don't have to read the test suite or project files to figure out what to name my solution file.

**Comment by Smarticles101** [comment]
This is something the java track has historically had issues with. We settled a while ago to [provide stubs on all exercises with difficulty 4 or less](https://github.com/exercism/java/blob/master/POLICIES.md#starter-implementations).

On the first exercise with difficulty 5 we have a hint in the readme explaining how to add the stubs and why they suddenly disappeared. We also keep the solution file structures fairly similar between exercise, making it easy for the student to look at previous solutions on how to structure the file.

I don't think this affects the auto-mentoring for java track too much, as each solution file only has one name that will compile successfully with the tests.

I like the thought of removing stubs progressively, or removing them at a certain difficulty level like the java track does.

**Comment by pgaspar** [comment]
> Having stubs enforces the file name of the solution, which reduces complexity for the auto-mentoring maintainers.

Aren't students able to split their work in multiple files and submit them all, though? I'm not sure if the analyzers will be able to assume there's only one file 🤔

I agree with this change specially for earlier exercises. I experienced it as an Elixir beginner in the Elixir track and I appreciated the confidence the stub gave me on that initial exercise.

**Comment by SleeplessByte** [comment]
> I'm not sure if the analyzers will be able to assume there's only one file 🤔

They probably will be `refer_to_mentor` for now.

**Comment by NobbZ** [comment]
> Aren't students able to split their work in multiple files and submit them all, though?

Yeah, probably, but some tracks make this easier then others, as well as language tooling might influence this ability.

In Erlang we have a strict relationship between module- and filenames. In elixir the name of the script under test is hard coded in the testsuite, the same is true for the bash suite. Probably even for many other languages as well. And so far I have not yet seen any exercise that would require to split implementation into multiple files (except for those languages that have a header file concept)

**Comment by petertseng** [comment]
Someone who has an abundance of time (wishful thinking on my part) would be able to read the linked issues from https://github.com/exercism/discussions/issues/114 and summarise any findings they get. Here is what I recall from that issue:

In general, we can see that there was a pattern that most tracks that had a discussion tended to decide in favour of having stub files. By nature of the survey method, there is no data for tracks that did not have a discussion on the record on GitHub.

From this past discussion, I retrieve one point in favour of stubs and contribute it to this discussion:

**Reduced annoyance for students**: We have a quote that creating the files is [supremely annoying](https://github.com/exercism/rust/issues/200#issuecomment-246203023). Especially true for tracks where the file needs to go in some deep directory structure.

Subnote on reduced annoyance applicable for statically-typed languages in an exercise that asks the student to implement > 1 function: Typically the entire test file must compile before it can be run; choosing not to provide a stub file means the student has to write this stub for all tested functions, rather than working on one at a time.

**Comment by iHiD** [comment]
Super helpful. Thank you, @petertseng. I'll add that link to my original post.

**Comment by kenliu** [comment]
As a student, I find it annoying as well (I brought this up @iHiD in Slack). Maybe not _supremely annoying_, but it seems like an unnecessary chore to have to do it on every single exercise.

While it doesn't take much time or effort to create a new file, if you consider that the ruby track has 90+ exercises, that a lot of wasted effort. Time is scarce for students, and we should find ways to help them use their time effectively.

**Comment by macta** [comment]
Can I step in a give a counter example: In Pharo I don't want to give a stub - the reason being that coming from  the source of TDD - the idea is that when you hit an error your environment helps you correct it. So the Smalltalk way is to run the first test - hit an error (the model your test references is not defined) - and then click on the "correct/create" button and the class is defined and execution continues... the next error being - the method you called is not defined, again click the create button and boom - new method and the debugger stops on the error "implementation not defined".

I think the intent is still the same - the student needs to be productive and start writing code as quickly as they can, but I think specifying the implementation isn't always the correct way. Of course in many languages - a template is the right way, but not all languages.

**Comment by SleeplessByte** [comment]
@iHiD cc @kytrinyx question regarding updating exercise to latest version.

When a student updates their exercise; how does it determine which files to overwrite and which files to leave alone?

**Comment by rpottsoh** [comment]
I'm not using stubs right now so this has never crossed my mind.  I'm very interested to know the answer too.

**Comment by NobbZ** [comment]
> When a student updates their exercise; how does it determine which files to overwrite and which files to leave alone?

From my observations, user submitted files are kept as they have been submitted, everything else gets synced with upstream.

This is annoying when the name of the file you are supposed to implement changes for some reason...

**Comment by SleeplessByte** [comment]
> From my observations, user submitted files are kept as they have been submitted, everything else gets synced with upstream.

Which means that if they have added tests, they will never get new tests.
Which also means that if we add stubs it **will not** overwrite their submissions.

**Comment by iHiD** [comment]
> From my observations, user submitted files are kept as they have been submitted, everything else gets synced with upstream.
> Which means that if they have added tests, they will never get new tests.
> Which also means that if we add stubs it will not overwrite their submissions.

All these points are correct.

**Comment by macta** [comment]
When students add tests, do you not ask them to create a separate TestCase? That’s what I want students to do in Pharo (but you’ve reminded me I need to make that more explicit in our track docs). The reasoning is that the separate test case then gets uploaded as part of their submission for review.

**Comment by NobbZ** [comment]
As most of us are using filedependant languages, and that its much simpler for students to just add a single function/test description to the existing file, rather than to create a complete new file which has to obey naming rules to be found be the testrunner while also having to include boilerplate code into this new file to set up the testing library and or doing before/after hooks for the tests, and and and…

**Comment by NobbZ** [comment]
> Which means that if they have added tests, they will never get new tests.

This is true regardless if you have stubs or not.
